File: worldcup-predictor/frontend/src/types/arbitrage.ts
/** M10 arbitrage types mirroring the Java DTOs. */

export type ArbMarketType = '1x2' | 'over_under' | 'asian_handicap' | 'btts';
export type ArbStatus = 'active' | 'expired' | 'stale';

export interface ArbBestQuote {
  odds: number | string;
  bookmaker: string;
  captured_at?: string | null;
  capturedAt?: string | null;
}

export interface ArbOpportunity {
  id: number;
  match_id?: number;
  matchId?: number;
  market_type?: ArbMarketType | string;
  marketType?: ArbMarketType | string;
  detected_at?: string;
  detectedAt?: string;
  arb_total?: number | string;
  arbTotal?: number | string;
  profit_margin?: number | string;
  profitMargin?: number | string;
  best_odds?: Record<string, ArbBestQuote>;
  bestOdds?: Record<string, ArbBestQuote>;
  stake_distribution?: Record<string, number | string>;
  stakeDistribution?: Record<string, number | string>;
  status: ArbStatus;
  expired_at?: string | null;
  expiredAt?: string | null;
}

export interface WatchlistEntry {
  id: number;
  userId: number;
  competitionId: number | null;
  marketTypes: string[] | null;
  minProfitMargin: number;
  notifyEnabled: boolean;
  createdAt: string;
}

export interface CreateWatchlistRequest {
  competitionId?: number | null;
  marketTypes?: string[] | null;
  minProfitMargin?: number;
  notifyEnabled?: boolean;
}

File: worldcup-predictor/src/api/schemas/arbitrage.py
"""Pydantic schemas for the M10 arbitrage API."""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ArbStatus = Literal["active", "expired", "stale"]
MarketType = Literal["1x2", "over_under", "asian_handicap", "btts"]


class ArbBestQuote(BaseModel):
    odds: Decimal
    bookmaker: str
    captured_at: str | None = None


class ArbOpportunityResponse(BaseModel):
    """Single arb opportunity row + stake math."""

    model_config = ConfigDict(extra="forbid")

    id: int
    match_id: int
    market_type: MarketType
    detected_at: datetime
    arb_total: Decimal
    profit_margin: Decimal
    best_odds: dict[str, ArbBestQuote]
    stake_distribution: dict[str, Decimal]
    status: ArbStatus
    expired_at: datetime | None = None


class CreateWatchlistRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    competition_id: int | None = None
    market_types: list[MarketType] | None = None
    min_profit_margin: Decimal = Field(default=Decimal("0.01"), ge=Decimal("0"))
    notify_enabled: bool = True


class WatchlistResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int
    user_id: int
    competition_id: int | None
    market_types: list[str] | None
    min_profit_margin: Decimal
    notify_enabled: bool
    created_at: datetime
